src/uu/ls/src/url_escaper.rs

```diff
@@ -0,0 +1,145 @@
+    }
+}
+
+pub struct UrlEscaper {
+    table: Box<[AsciiOrEscape; 256]>,
+}
+
+impl UrlEscaper {
+    pub fn new() -> Self {
+        let mut table = Vec::with_capacity(256);
+        static HEX_BYTES: &[u8; 16] = b"0123456789abcdef";
+        let hex_bytes = AsciiStr::from_ascii(HEX_BYTES).unwrap();
+        for &c0 in hex_bytes {
+            for &c1 in hex_bytes {
+                table.push(AsciiOrEscape::escape(c0, c1));
+            }
+        }
+        assert_eq!(table.len(), 256);
+        for r in [b'0'..=b'9', b'A'..=b'Z', b'a'..=b'z'] {
+            for b in r {
+                table[b as usize] = AsciiOrEscape::ascii(AsciiChar::from_ascii(b).unwrap());
+            }
+        }
+        for b in [b'_', b'-', b'.', b'~'] {
+            table[b as usize] = AsciiOrEscape::ascii(AsciiChar::from_ascii(b).unwrap());
+        }
+        Self {
+            table: table.into_boxed_slice().try_into().map_err(|_| ()).unwrap(),
+        }
+    }
+
+    pub fn into_path_escaper(mut self) -> Self {
+        let slash = AsciiOrEscape::ascii(AsciiChar::from_ascii(b'/').unwrap());
+        self.table[b'/' as usize] = slash;
+        if std::path::MAIN_SEPARATOR != '/' {
+            if let Some(main_separator) = self.table.get_mut(std::path::MAIN_SEPARATOR as usize) {
+                *main_separator = slash;
+            }
+        }
+        self
+    }
+
+    pub fn escape<'a>(&self, s: &'a [u8]) -> Cow<'a, AsciiStr> {
+        let table = &*self.table;
+
+        let mut not_escapes = 0usize;
+        let mut not_identities = false;
+        for b in s.iter().copied() {
+            if std::path::MAIN_SEPARATOR != '/' && (std::path::MAIN_SEPARATOR as usize) < 256 {
+                not_identities |= b == (std::path::MAIN_SEPARATOR as u8);
+            }
+            // cannot overflow because it will be at most s.len() since we are summing 0 or 1 values
+            not_escapes = not_escapes.wrapping_add(table[b as usize].is_byte() as usize);
+        }
+
+        if not_escapes == s.len() && !not_identities {
+            // SAFETY: this is safe because the string only has ASCII characters (otherwise it would need escapes)
+            Cow::Borrowed(unsafe { AsciiStr::from_ascii_unchecked(s) })
+        } else {
+            static ERROR: &str = "escaped string length too large to fit in memory";
+            // Vec only supports capacity up to isize::MAX, so we must use isize to do all checks
+            let len = isize::try_from(s.len()).expect(ERROR);
+            // can't wrap because not_escapes <= s.len
+            let escapes = len.wrapping_sub(not_escapes as isize);
+            // must check for overflow since we are using unsafe code
+            let size = len
+                .checked_add(escapes)
+                .expect(ERROR)
+                .checked_add(escapes)
+                .expect(ERROR);
+
+            let mut res = Vec::with_capacity(size as usize);
+            let mut out: *mut AsciiChar = res.as_mut_ptr();
```

Comment:
I still would like to see this implemented with a slice, or at least some information about how much faster this is, so we'll be able to judge whether the use of raw pointers is justified.